samples/sid_end_device/src/hello/app.c

```diff
@@ -213,51 +216,105 @@ static void free_sid_hello_event_ctx(void *ctx)
+	hello->msg.data = sid_hal_malloc(hello->msg.size);
+	if (!hello->msg.data) {
+		sid_hal_free(hello);
+		LOG_ERR("Failed to allocate memory for message data");
+		return;
+	}
+	memcpy(hello->msg.data, payload, hello->msg.size);
 
-		hello->desc.type = SID_MSG_TYPE_NOTIFY;
-		hello->desc.link_type = SID_LINK_TYPE_ANY;
-		hello->desc.link_mode = SID_LINK_MODE_CLOUD;
+	hello->desc.type = SID_MSG_TYPE_NOTIFY;
+	hello->desc.link_type = SID_LINK_TYPE_ANY;
+	hello->desc.link_mode = SID_LINK_MODE_CLOUD;
 
-		int err = sidewalk_event_send(SID_EVENT_SEND_MSG, hello, free_sid_hello_event_ctx);
+	int err = sidewalk_event_send(sidewalk_event_send_msg, hello, free_sid_hello_event_ctx);
+	if (err) {
+		free_sid_hello_event_ctx(hello);
+		LOG_ERR("Send event err %d", err);
+	} else {
+		application_state_sending(&global_state_notifier, true);
+	}
+}
+
+static void app_event_dfu_mode(sidewalk_ctx_t *sid, void *ctx)
+{
+	bool *enter_dfu_state = (bool *)ctx;
+	int err = -ENOTSUP;
+	if (*enter_dfu_state) {
+		// go to DFU state
+		sidewalk_event_exit(sid, NULL);
+		LOG_INF("Entering into DFU mode");
+#if defined(CONFIG_SIDEWALK_DFU_SERVICE_BLE)
+		err = nordic_dfu_ble_start();
+#endif
 		if (err) {
-			free_sid_hello_event_ctx(hello);
-			LOG_ERR("Send event err %d", err);
-		} else {
-			application_state_sending(&global_state_notifier, true);
+			LOG_ERR("dfu start err %d", err);
 		}
+	} else {
+		// Exit from DFU state
+#if defined(CONFIG_SIDEWALK_DFU_SERVICE_BLE)
+		err = nordic_dfu_ble_stop();
+#endif
+		if (err) {
+			LOG_ERR("dfu stop err %d", err);
+		}
+		sidewalk_event_autostart(sid, NULL);
+	}
+}
+
+static void app_btn_dfu_state(uint32_t unused)
+{
+	ARG_UNUSED(unused);
+	static bool go_to_dfu_state = true;
+	int e = sidewalk_event_send(app_event_dfu_mode, &go_to_dfu_state, NULL);
+	if (e) {
+		LOG_ERR("Send event err %d", e);
 		return;
 	}
+	application_state_working(&global_state_notifier, !go_to_dfu_state);
+	go_to_dfu_state = !go_to_dfu_state;
+}
+
+static void app_btn_connect(uint32_t unused)
+{
+	ARG_UNUSED(unused);
+	(void)sidewalk_event_send(sidewalk_event_connect, NULL, NULL);
+}
 
-	sidewalk_event_send((sidewalk_event_t)event, NULL, NULL);
+static void app_btn_factory_reset(uint32_t unused)
+{
+	ARG_UNUSED(unused);
+	(void)sidewalk_event_send(sidewalk_event_factory_reset, NULL, NULL);
+}
+
+static void app_btn_link_switch(uint32_t unused)
+{
+	ARG_UNUSED(unused);
+	(void)sidewalk_event_send(sidewalk_event_link_switch, NULL, NULL);
 }
 
 static int app_buttons_init(void)
 {
-	button_set_action_short_press(DK_BTN1, app_button_handler, SID_EVENT_SEND_MSG);
-	button_set_action_long_press(DK_BTN1, app_button_handler, SID_EVENT_NORDIC_DFU);
-	button_set_action_short_press(DK_BTN2, app_button_handler, SID_EVENT_CONNECT);
-	button_set_action_long_press(DK_BTN2, app_button_handler, SID_EVENT_FACTORY_RESET);
-	button_set_action(DK_BTN3, app_button_handler, SID_EVENT_LINK_SWITCH);
+	button_set_action_short_press(DK_BTN1, app_btn_send_msg, 0);
```

Comment:
minor: can add sth like `#define UNUSED_ARGS (0U)` to avoid magic numbers (just suggestion)